### ISSUE TITLE
[Mobile] Use NNAPI to throw exception ANEURALNETWORKS_BAD_DATA

### ISSUE THREAD

**Issue by cmelody999** [microsoft/onnxruntime#16224]
### Use NNAPI to throw exception ANEURALNETWORKS_BAD_DATA

We converted the Pytorch model of ResNet-50 to a ONNX model and the java code on Android phone uses version 1.14.0 of ONNX Runtime to enable NNAPI via ortOption.addNnapi() and reports the following error:
![image](https://github.com/microsoft/onnxruntime/assets/5425469/0926f422-d7b3-4274-8cec-f05c652d3b4a)
Caused by: ai.onnxruntime.OrtException: Error code - ORT_FAIL - message: model_builder.cc:457 AddOperandFromPersistMemoryBuffer ResultCode: ANEURALNETWORKS_BAD_DATA

How to get more details about the error and how to fix it? Thank you very much!

The model download address: https://github.com/cmelody999/model

Associated issues: https://github.com/microsoft/onnxruntime/issues/16194

### To reproduce

Convert the Pytorch model of ResNet-50(https://github.com/OFA-Sys/Chinese-CLIP)  to a ONNX model and enable NNAPI via ortOption.addNnapi().
![image](https://github.com/microsoft/onnxruntime/assets/5425469/630bca01-64ac-413f-b548-9ea81c610eaf)

### Platform

Android

### OS Version

android 12

### ONNX Runtime Installation

Built from Source

### Package Name (if 'Released Package')

None

### ONNX Runtime Version or Commit ID

1.14.0

### ONNX Runtime API

Java/Kotlin

### Architecture

X64

### Execution Provider

NNAPI

**Comment by cmelody999** [comment]
We removed the ReduceMean operator and replaced it with an operator that randomly initializes a Gaussian normal distribution vector. The similarity score gets worse, and turning on NNAPI still crashes the same, but running check_onnx_model_mobile_usability detects that the model should perform well with NNAPI as is: YES (1 partitions with a total (1 partitions with a total of 163/163 nodes can be handled by the NNAPI EP).

**Comment by wejoncy** [comment]
![image](https://github.com/microsoft/onnxruntime/assets/9417365/4b0eb10e-1313-4eaa-9def-c29dd951790b)
![image](https://github.com/microsoft/onnxruntime/assets/9417365/ea449b22-620a-42d4-ac30-168f19f241a3)

The last Gather Op, indice shouldn't be a constant scalar.

**Comment by skottmckay** [comment]
Whatever is responsible for the last Gather in the ONNX model is most likely using a scalar with value 0 for the indices instead of an array with a single entry of value 0. That Gather node is selecting the output for unnorm_image_features.

**Comment by wejoncy** [comment]
> 

NNAPI EP doesnt't handle scalar `indices`

**Comment by wejoncy** [comment]
> Whatever is responsible for the last Gather in the ONNX model is most likely using a scalar with value 0 for the indices instead of an array with a single entry of value 0. That Gather node is selecting the output for unnorm_image_features.

NNAPI EP doesnt't handle scalar `indices`

**Comment by skottmckay** [comment]
> NNAPI EP doesn't handle scalar `indices`

I was mainly suggesting it would be better to fix the pytorch model rather than hacking the exported onnx model. As it's only the very last Gather it should be relatively easy to narrow down where in the original model that is coming from.

**Comment by wejoncy** [comment]
> > NNAPI EP doesn't handle scalar `indices`
> 
> I was mainly suggesting it would be better to fix the pytorch model rather than hacking the exported onnx model. As it's only the very last Gather it should be relatively easy to narrow down where in the original model that is coming from.

According to the ONNX spec, Gather `indices` is an array instead of scalar, it looks like a bug of exporter?

**Comment by skottmckay** [comment]
Yeah - I don't think the exporter should emit `indices` as a scalar. That said, the ONNX spec has always been a little unclear about how/when a scalar (rank 0 with shape {}) and a single 1D value (rank 1 with shape {1}) are considered equivalent or interchangable.

Additionally the Reshape looks wrong based on symbolic_shape_infer output, but ironically that makes the ONNX shape inferencing pass. There's a Reshape to 50 x 1 x 1024 followed by the Gather with scalar indices and an output shape specified in the model of 1 x 1024 (i.e. the output shape is not inferred as it's a model output). Technically that is the correct shape as q==0 according to the [Gather spec](https://github.com/onnx/onnx/blob/main/docs/Operators.md#Gather) with the scalar indices, so a 3D input with 2D output is 'correct'. AFAIK that wouldn't be an expected usage of Gather though as it's really a Gather + Squeeze.

I would think the more correct converter output would be a Reshape to 50 x 1024 and a Gather with 1D indices, but I don't know what the pytorch model code for that part of the conversion looked like to say how much that contributed to the issue.


<img width="213" alt="image" src="https://github.com/microsoft/onnxruntime/assets/979079/b30ea766-e068-4575-9e70-055304496fcd">

**Comment by cmelody999** [comment]
![img_v2_18b7e9d9-ac77-4825-b1be-3c8bf4a0a23g](https://github.com/microsoft/onnxruntime/assets/5425469/ef7676f1-0e28-405f-a08e-81b36bb1b82f)
There is no problem with the onnx model on pc, and in the screenshot is the definition of the indices of the gather operator.
How to modify this constant_252 operator?

**Comment by wejoncy** [comment]
> ![img_v2_18b7e9d9-ac77-4825-b1be-3c8bf4a0a23g](https://user-images.githubusercontent.com/5425469/243996460-ef7676f1-0e28-405f-a08e-81b36bb1b82f.jpg) There is no problem with the onnx model on pc, and in the screenshot is the definition of the indices of the gather operator. How to modify this constant_252 operator?

My code pasted above should workaround this issue temprorary.

**Comment by wejoncy** [comment]
> Yeah - I don't think the exporter should emit `indices` as a scalar. That said, the ONNX spec has always been a little unclear about how/when a scalar (rank 0 with shape {}) and a single 1D value (rank 1 with shape {1}) are considered equivalent or interchangable.
> 
> Additionally the Reshape looks wrong based on symbolic_shape_infer output, but ironically that makes the ONNX shape inferencing pass. There's a Reshape to 50 x 1 x 1024 followed by the Gather with scalar indices and an output shape specified in the model of 1 x 1024 (i.e. the output shape is not inferred as it's a model output). Technically that is the correct shape as q==0 according to the [Gather spec](https://github.com/onnx/onnx/blob/main/docs/Operators.md#Gather) with the scalar indices, so a 3D input with 2D output is 'correct'. AFAIK that wouldn't be an expected usage of Gather though as it's really a Gather + Squeeze.
> 
> I would think the more correct converter output would be a Reshape to 50 x 1024 and a Gather with 1D indices, but I don't know what the pytorch model code for that part of the conversion looked like to say how much that contributed to the issue.
> 
> <img alt="image" width="213" src="https://user-images.githubusercontent.com/979079/243976898-b30ea766-e068-4575-9e70-055304496fcd.png">

It seems quite reasonable in the code for the last `Gather`.

https://github.com/OFA-Sys/Chinese-CLIP/blob/2c38d03557e50eadc72972b272cebf840dbc34ea/cn_clip/clip/model.py#LL103C16-L104C1

**Comment by wejoncy** [comment]
> If the `x[0]` is a resulting in the Reshape + Gather it doesn't seem like the best option as the alternative would be to not add the Reshape from 50 x 1024 to 50 x 1 x 1024 and to instead do the Gather with indices shape {1} on the output from the GEMM.
> 
> If we think this usage of Gather is a valid pattern we could try and support it in NNAPI by converting the scalar indices value to a 1D input and add a Squeeze of `axis` to remove that dim from the output.


`x[0]` is equivant of  Gather with scalar indices.
`x[[0]]` is resulting Gather + tensor indices.
`Reshape` is the result of `multi_head_attention_forward`. 
It's easy to be reproduced by 
```
from torch import nn

class AttentionPool2d(nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, x):
        return x[[0]]

model = AttentionPool2d().cuda()
input_names = [ "x" ]
output_names = [ "output1" ]
dummy_input = torch.randn(1, 3, 8, 8, device="cuda")
torch.onnx.export(model, dummy_input, "att.onnx", verbose=True, opset_version=14, input_names=input_names, output_names=output_names)
```

Not sure if it's expected in onnx exporter.

**Comment by cmelody999** [comment]
Model inference time consumption before NNAPI is not enabled:

**Comment by skottmckay** [comment]
This is expected. Without the ReduceMean you have the cost of transferring from NPU back to CPU back to NPU in the middle of the model.

Also note that NNAPI performance is extremely device specific. The low-level implementation of the internal NNAPI API is done by the hardware vendors. If they have a poor implementation performance can be worse than CPU. If the hardware vendor has not implemented a specific operator NNAPI will fall back to a reference CPU implementation (i.e. the simplest implementation possible which is not optimized in any way). So even with the ReduceMean it's possible that on some devices NNAPI performance may be worse than CPU.
